File: plugins/mobile-link/lib/dlp.js
/**
 * Pure DLP v1 helpers: frame codec, stream-id mapping, waterfall dedupe and
 * reconnect backoff. No sockets, no timers, no filesystem — everything here is
 * driven by `node --test` without a DSH instance.
 *
 * Wire contract: docs/RELAY-PROTOCOL.md §3.
 */

/** Frame ceiling aligned with DSH's image-attachment limit (spec §5). */
export const MAX_FRAME_BYTES = 32 * 1024 * 1024

export const DEVICE_TO_AGENT = new Set(['req', 'open', 'cancel', 'eventResult', 'ping', 'hello'])
export const AGENT_TO_DEVICE = new Set(['res', 'item', 'end', 'streamError', 'event', 'hostStatus', 'pong', 'error'])
/** Relay-internal control frames carrying device lifecycle (see relay/NOTES.md). */
export const RELAY_CONTROL = new Set(['deviceAttach', 'deviceDetach'])

const ID_FRAMES = new Set(['req', 'open', 'cancel', 'res', 'item', 'end', 'streamError', 'eventResult'])

export function encodeFrame(frame) {
  return JSON.stringify(frame)
}

/**
 * Decode one wire payload.
 * @returns the frame object, or `undefined` when the payload is unusable
 *          (not JSON, not an object, or missing a non-empty string `t`).
 */
export function decodeFrame(raw) {
  if (typeof raw !== 'string') return undefined
  if (raw.length > MAX_FRAME_BYTES) return undefined
  let value
  try {
    value = JSON.parse(raw)
  } catch {
    return undefined
  }
  if (value === null || typeof value !== 'object' || Array.isArray(value)) return undefined
  if (typeof value.t !== 'string' || value.t.length === 0) return undefined
  return value
}

export function frameType(frame) {
  return typeof frame?.t === 'string' ? frame.t : ''
}

/** Structural check for one device -> agent frame; returns an error string or undefined. */
export function deviceFrameError(frame) {
  const kind = frameType(frame)
  if (!DEVICE_TO_AGENT.has(kind)) return undefined
  if (ID_FRAMES.has(kind) && !nonEmptyString(frame.id)) return `${kind}: missing \`id\``
  if (kind === 'req' && !nonEmptyString(frame.method)) return 'req: missing `method`'
  if (kind === 'open' && !nonEmptyString(frame.endpoint)) return 'open: missing `endpoint`'
  return undefined
}

export function nonEmptyString(value) {
  return typeof value === 'string' && value.length > 0
}

export function deviceIdOf(frame) {
  return nonEmptyString(frame?.deviceId) ? frame.deviceId : undefined
}

/**
 * Exponential backoff with ±jitter, capped (spec §4.2: 1s -> 2s -> ... -> 30s ±20%).
 * @param attempt zero-based retry index.
 */
export function nextBackoff(attempt, { baseMs = 1000, maxMs = 30000, jitter = 0.2, random = Math.random } = {}) {
  const exponent = Math.max(0, Math.min(30, Number(attempt) || 0))
  const raw = Math.min(maxMs, baseMs * 2 ** exponent)
  const spread = raw * jitter
  const offset = (random() * 2 - 1) * spread
  return Math.max(0, Math.round(raw + offset))
}

/**
 * Split a relay URL into its HTTP and WebSocket bases, keeping the path prefix.
 *
 * The relay is published under a path on an existing site, so the configured
 * base carries a prefix (`wss://relay.example.com/dsh-link`) that must be
 * *kept* when an endpoint is appended — never replaced. A trailing slash is
 * tolerated and normalised away. Accepts `wss://host`, `https://host/prefix`,
 * `ws://127.0.0.1:8787` and bare hosts.
 */
export function normalizeRelayUrl(raw) {
  const text = String(raw ?? '').trim()
  if (!text) throw new Error('relay URL is required')
  const withScheme = /^[a-z][a-z0-9+.-]*:\/\//i.test(text) ? text : `wss://${text}`
  const url = new URL(withScheme)
  const secure = url.protocol === 'wss:' || url.protocol === 'https:'
  if (!['ws:', 'wss:', 'http:', 'https:'].includes(url.protocol)) {
    throw new Error(`unsupported relay scheme ${url.protocol}`)
  }
  const authority = `${url.host}`
  const basePath = url.pathname.replace(/\/+$/, '')
  return {
    httpBase: `${secure ? 'https' : 'http'}://${authority}${basePath}`,
    wsBase: `${secure ? 'wss' : 'ws'}://${authority}${basePath}`,
    secure,
    authority,
  }
}

/**
 * The relay address the repository ships as a placeholder.
 *
 * It is a *sentinel*, not a deployment: a value equal to it means "no relay has
 * been configured here", never "dial relay.example.com". The distinction is
 * load-bearing — a configured placeholder used to outrank the identity already
 * enrolled in `agent.json`, so one DSH restart pointed an enrolled computer at
 * the placeholder *and* wrote that back over the real address, which left the
 * phone unable to connect until somebody noticed. Anything that resolves an
 * address has to recognise the sentinel.
 */
export const PLACEHOLDER_RELAY_URL = 'wss://relay.example.com/dsh-link'

/**
 * True when `raw` is, or normalises to, the repository placeholder — a trailing
 * slash or an `https://` spelling of the same address still counts.
 */
export function isPlaceholderRelayUrl(raw) {
  if (typeof raw !== 'string' || !raw.trim()) return false
  try {
    return normalizeRelayUrl(raw).wsBase === normalizeRelayUrl(PLACEHOLDER_RELAY_URL).wsBase
  } catch {
    return false
  }
}

/**
 * Join one endpoint onto a relay base without losing the base's path prefix.
 * `('wss://host/prefix/', '/link/agent')` -> `'wss://host/prefix/link/agent'`.
 */
export function joinRelayPath(base, suffix) {
  const head = String(base ?? '').replace(/\/+$/, '')
  const tail = String(suffix ?? '').replace(/^\/+/, '')
  return `${head}/${tail}`
}

/** The connector's relay endpoint, keeping any configured path prefix. */
export function agentEndpoint(relayUrl) {
  return joinRelayPath(normalizeRelayUrl(relayUrl).wsBase, 'link/agent')
}

/** The relay's pairing-code endpoint, keeping any configured path prefix. */
export function pairCodeEndpoint(relayUrl) {
  return joinRelayPath(normalizeRelayUrl(relayUrl).httpBase, 'pair/code')
}

/** The relay's self-service enrollment endpoint (`POST /agents/enroll`). */
export function enrollEndpoint(relayUrl) {
  return joinRelayPath(normalizeRelayUrl(relayUrl).httpBase, 'agents/enroll')
}

/**
 * The `dsh://pair?relay=<relay>&code=<code>` deep link the desktop shows and
 * the iOS app scans. `relay` keeps the full path prefix so the phone can reach
 * `/pair/claim` and `/link/device` under exactly the same base.
 */
export function qrPayload({ relay, code }) {
  const params = new URLSearchParams({ relay: String(relay), code: String(code) })
  return `dsh://pair?${params.toString()}`
}

/**
 * DLP id <-> mux streamId mapping. One mux stream per (device, DLP stream) so
 * two devices can both call their stream `"2"` without colliding.
 */
export class StreamTable {
  constructor() {
    /** @type {Map<string, {deviceId: string, dlpId: string, muxId: string, endpoint: string}>} */
    this.byMux = new Map()
    /** @type {Map<string, string>} `${deviceId}\u0000${dlpId}` -> muxId */
    this.byDlp = new Map()
  }

  static key(deviceId, dlpId) {
    return `${deviceId}\u0000${dlpId}`
  }

  /** @returns {boolean} false when this device already owns that DLP id. */
  add({ deviceId, dlpId, muxId, endpoint }) {
    if (this.byDlp.has(StreamTable.key(deviceId, dlpId))) return false
    this.byMux.set(muxId, { deviceId, dlpId, muxId, endpoint })
    this.byDlp.set(StreamTable.key(deviceId, dlpId), muxId)
    return true
  }

  muxOf(deviceId, dlpId) {
    return this.byDlp.get(StreamTable.key(deviceId, dlpId))
  }

  entryOfMux(muxId) {
    return this.byMux.get(muxId)
  }

  /** Remove one (device, DLP) pair; returns its mux streamId when it existed. */
  removeByDlp(deviceId, dlpId) {
    const key = StreamTable.key(deviceId, dlpId)
    const muxId = this.byDlp.get(key)
    if (muxId === undefined) return undefined
    this.byDlp.delete(key)
    this.byMux.delete(muxId)
    return muxId
  }

  /** Remove one mux stream; returns its entry when it existed. */
  removeByMux(muxId) {
    const entry = this.byMux.get(muxId)
    if (entry === undefined) return undefined
    this.byMux.delete(muxId)
    this.byDlp.delete(StreamTable.key(entry.deviceId, entry.dlpId))
    return entry
  }

  /** Every mux streamId owned by one device (used when it disconnects). */
  removeDevice(deviceId) {
    const removed = []
    for (const entry of [...this.byMux.values()]) {
      if (entry.deviceId === deviceId) {
        removed.push(entry.muxId)
        this.byMux.delete(entry.muxId)
        this.byDlp.delete(StreamTable.key(entry.deviceId, entry.dlpId))
      }
    }
    return removed
  }

  get size() {
    return this.byMux.size
  }
}

/**
 * First-answer-wins bookkeeping for `$events` waterfalls (spec §4.1.6).
 *
 * `claim` reserves an event synchronously so two devices answering at the same
 * moment cannot both reach `$events/result`; `release` hands it back when the
 * reservation failed to produce a delivered answer.
 */
export class WaterfallDedupe {
  constructor({ ttlMs = 5 * 60 * 1000, now = Date.now } = {}) {
    this.ttlMs = ttlMs
    this.now = now
    /** @type {Map<string, number>} eventId -> claimedAt */
    this.claimed = new Map()
  }

  /** @returns {boolean} true when this caller won the right to answer. */
  claim(eventId) {
    if (!nonEmptyString(eventId)) return false
    this.prune()
    if (this.claimed.has(eventId)) return false
    this.claimed.set(eventId, this.now())
    return true
  }

  release(eventId) {
    this.claimed.delete(eventId)
  }

  has(eventId) {
    return this.claimed.has(eventId)
  }

  prune(at = this.now()) {
    for (const [eventId, claimedAt] of this.claimed) {
      if (at - claimedAt > this.ttlMs) this.claimed.delete(eventId)
    }
  }

  get size() {
    return this.claimed.size
  }
}

/** `{t:'pong', ts}` for a device/relay heartbeat, preserving `ts` verbatim. */
export function pongFor(frame) {
  return { t: 'pong', ts: frame?.ts }
}

/** `{t:'res', ...}` mapped from a DSH `RemoteResult` (`{ok,value}|{ok:false,error}`). */
export function resultFrame(dlpId, result, deviceId) {
  const base = { t: 'res', id: dlpId, deviceId }
  if (result && result.ok === true) return { ...base, ok: true, value: result.value }
  const error = result?.error
  return {
    ...base,
    ok: false,
    error: {
      code: nonEmptyString(error?.code) ? error.code : 'gateway/failed',
      message: nonEmptyString(error?.message) ? error.message : 'DSH request failed',
      details: error && typeof error.details === 'object' && error.details !== null ? error.details : {},
    },
  }
}

File: relay/dlp.py
"""DLP v1 frame codec and validation.

The relay is a *transparent* forwarder: it never interprets DSH semantics. It
only needs enough structure to know which direction a frame travels and which
device it belongs to. Everything else is passed through verbatim, including
frame types this version does not know (``t`` is forward-compatible by design).

Wire summary (see ``docs/RELAY-PROTOCOL.md`` §3):

  device -> agent   req / open / cancel / eventResult / ping / hello
  agent  -> device  res / item / end / streamError / event / hostStatus / pong / error

Two frame families are relay-internal control frames (they never cross to the
other side): ``deviceAttach`` / ``deviceDetach`` (relay -> agent) describe a
device's lifecycle so the agent can open and tear down that device's ``$events``
streams. The relay also stamps ``deviceId`` onto every forwarded frame, because
one agent WebSocket multiplexes every device.
"""

from __future__ import annotations

import json
from typing import Any

PROTOCOL_VERSION = 1

#: Hard frame ceiling, aligned with DSH's image-attachment limit.
MAX_FRAME_BYTES = 32 * 1024 * 1024

DEVICE_TO_AGENT = frozenset({"req", "open", "cancel", "eventResult", "ping", "hello"})
AGENT_TO_DEVICE = frozenset({"res", "item", "end", "streamError", "event", "hostStatus", "pong", "error"})
RELAY_TO_AGENT = frozenset({"deviceAttach", "deviceDetach"})

#: Frame types that carry a unary request and therefore need an ``id``.
_ID_FRAMES = frozenset({"req", "open", "cancel", "res", "item", "end", "streamError", "eventResult"})


class FrameError(ValueError):
    """A frame that cannot be routed (not JSON, not an object, no usable ``t``)."""


def parse_frame(raw: str | bytes) -> dict[str, Any]:
    """Decode one WebSocket text payload into a frame object.

    Raises :class:`FrameError` for anything that is not a JSON object with a
    non-empty string ``t``. Unknown ``t`` values are *not* an error here; the
    caller decides whether to forward or ignore them.
    """
    if isinstance(raw, (bytes, bytearray, memoryview)):
        try:
            text = bytes(raw).decode("utf-8")
        except UnicodeDecodeError as error:
            raise FrameError("frame is not valid UTF-8") from error
    else:
        text = raw
    if len(text) > MAX_FRAME_BYTES:
        raise FrameError(f"frame exceeds {MAX_FRAME_BYTES} bytes")
    try:
        value = json.loads(text)
    except (ValueError, TypeError) as error:
        raise FrameError("frame is not JSON") from error
    if not isinstance(value, dict):
        raise FrameError("frame must be a JSON object")
    frame_type = value.get("t")
    if not isinstance(frame_type, str) or not frame_type:
        raise FrameError("frame is missing a non-empty string `t`")
    return value


def encode_frame(frame: dict[str, Any]) -> str:
    """Encode a frame. ``ensure_ascii`` keeps the payload byte-exact across hops."""
    return json.dumps(frame, ensure_ascii=False, separators=(",", ":"))


def frame_type(frame: dict[str, Any]) -> str:
    return str(frame.get("t"))


def has_id(frame: dict[str, Any]) -> bool:
    value = frame.get("id")
    return isinstance(value, str) and len(value) > 0


def validate_from_device(frame: dict[str, Any]) -> str | None:
    """Return an error string when a device frame is structurally unusable."""
    kind = frame_type(frame)
    if kind in DEVICE_TO_AGENT and kind in _ID_FRAMES and not has_id(frame):
        return f"{kind}: missing `id`"
    if kind == "req" and not isinstance(frame.get("method"), str):
        return "req: missing `method`"
    if kind == "open" and not isinstance(frame.get("endpoint"), str):
        return "open: missing `endpoint`"
    return None


def validate_from_agent(frame: dict[str, Any]) -> str | None:
    """Return an error string when an agent frame is structurally unusable."""
    kind = frame_type(frame)
    if kind in AGENT_TO_DEVICE and kind in _ID_FRAMES and not has_id(frame):
        return f"{kind}: missing `id`"
    return None


def device_id_of(frame: dict[str, Any]) -> str | None:
    value = frame.get("deviceId")
    return value if isinstance(value, str) and value else None


def addressing_device(frame: dict[str, Any]) -> dict[str, Any] | None:
    """Return a copy addressed at one device, dropping the relay-internal key."""
    device_id = device_id_of(frame)
    if device_id is None:
        return None
    addressed = dict(frame)
    addressed.pop("deviceId", None)
    addressed["deviceId"] = device_id
    return addressed


def strip_device_id(frame: dict[str, Any]) -> dict[str, Any]:
    """Copy of ``frame`` without the relay-internal ``deviceId`` key."""
    if "deviceId" not in frame:
        return frame
    out = dict(frame)
    out.pop("deviceId", None)
    return out


def host_status(*, online: bool, agent_id: str | None = None, name: str | None = None,
                version: str | None = None, extra: dict[str, Any] | None = None) -> dict[str, Any]:
    """Build the ``hostStatus`` frame the relay pushes to devices."""
    info: dict[str, Any] = {"online": online}
    if agent_id is not None:
        info["agentId"] = agent_id
    if name is not None:
        info["name"] = name
    if version is not None:
        info["version"] = version
    if extra:
        info.update(extra)
    return {"t": "hostStatus", "info": info}


def error_frame(code: str, message: str, *, fatal: bool = False,
                details: dict[str, Any] | None = None) -> dict[str, Any]:
    frame: dict[str, Any] = {"t": "error", "code": code, "message": message}
    if fatal:
        frame["fatal"] = True
    if details:
        frame["details"] = details
    return frame


def device_attach_frame(device_id: str, *, name: str | None = None,
                        model: str | None = None, agent_id: str | None = None) -> dict[str, Any]:
    frame: dict[str, Any] = {"t": "deviceAttach", "deviceId": device_id}
    info: dict[str, Any] = {}
    if name is not None:
        info["name"] = name
    if model is not None:
        info["model"] = model
    if agent_id is not None:
        info["agentId"] = agent_id
    if info:
        frame["device"] = info
    return frame


def device_detach_frame(device_id: str, *, reason: str | None = None) -> dict[str, Any]:
    frame: dict[str, Any] = {"t": "deviceDetach", "deviceId": device_id}
    if reason:
        frame["reason"] = reason
    return frame
